src/plugins/dashboard/public/dashboard_container/state/diffing/dashboard_diffing_integration.ts

```diff
@@ -118,18 +118,17 @@ export function startDiffingDashboardState(
     startWith(null),
     debounceTime(CHANGE_CHECK_DEBOUNCE),
     switchMap(() => {
-      return new Observable<Partial<DashboardContainerInput>>((observer) => {
+      return (async () => {
         const {
           explicitInput: currentInput,
           componentState: { lastSavedInput },
         } = this.getState();
-        getDashboardUnsavedChanges
-          .bind(this)(lastSavedInput, currentInput)
-          .then((unsavedChanges) => {
-            if (observer.closed) return;
-            observer.next(unsavedChanges);
-          });
-      });
+        const unsavedChanges = await getDashboardUnsavedChanges.bind(this)(
+          lastSavedInput,
+          currentInput
+        );
+        return unsavedChanges;
+      })();
```

Comment:
Switch to return a promise rather than an observable - this is a stylistic change and has nothing to do with the actual bug.

Reply:
Thanks for making this change. IMO it's much more obvious what's actually happening here when we use a promise